src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/eg_mqtt_broker_client.c

```diff
@@ -0,0 +1,157 @@
+/**
+ * @file adps_gen2.c
+ * @brief Implements the Azure DPS Gen2 communication utility.
+ *
+ * @copyright Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT License.
+ */
+#include "aduc/mqtt_client.h"
+#include "aduc/ADUC_MQTT_Client_client.h"
+#include "aduc/config_utils.h"
+#include "aduc/logging.h"
+#include "aduc/string_c_utils.h" // ADUC_StringFormat
+#include "stdlib.h" // memset
+
+// Default MQTT protocol version for Event Grid MQTT Broker is v5 (5)
+#define DEFAULT_EG_MQTT_PROTOCOL_VERSION 5
+
+/**
+ * @brief Read the MQTT broker connection data from the config file.
+ *
+ * @remark The caller is responsible for freeing the returned settings by calling FreeMqttBrokerSettings function.
+*/
+bool ReadMqttBrokerSettings(ADUC_MQTT_SETTINGS* settings)
+{
+    bool result = false;
+    const ADUC_ConfigInfo* config = NULL;
+    const ADUC_AgentInfo* agent_info = NULL;
+
+    if (settings == NULL)
+    {
+        goto done;
+    }
+
+    memset(settings, 0, sizeof(*settings));
+
+    config = ADUC_ConfigInfo_GetInstance();
+    if (config == NULL)
+    {
+        goto done;
+    }
+
+    agent_info = ADUC_ConfigInfo_GetAgent(config, 0);
+
+    // Currently only support 'ADPS2/MQTT' connection data.
+    if (strcmp(agent_info->connectionType, ADUC_CONNECTION_TYPE_MQTTBROKER) != 0)
+    {
+        goto done;
+    }
+
+    // Read the x.506 certificate authentication data.
+    if (!ADUC_AgentInfo_ConnectionData_GetStringField(agent_info, "mqttBroker.caFile", &settings->caFile))
+    {
+        settings->caFile = NULL;
+    }
+
+    if (!ADUC_AgentInfo_ConnectionData_GetStringField(agent_info, "mqttBroker.certFile", &settings->certFile))
+    {
+        settings->certFile = NULL;
+    }
+
+    if (!ADUC_AgentInfo_ConnectionData_GetStringField(agent_info, "mqttBroker.keyFile", &settings->keyFile))
+    {
+        settings->keyFile = NULL;
+    }
+
+    if (!ADUC_AgentInfo_ConnectionData_GetStringField(
+            agent_info, "mqttBroker.keyFilePassword", &settings->keyFilePassword))
+    {
+        settings->keyFilePassword = NULL;
+    }
+
+
+    // TODO (nox-msft) - For MQTT broker, userName must match the device Id (usually the CN of the device's certificate)
+    if (IsNullOrEmpty(settings->username))
+    {
+        if (!ADUC_AgentInfo_ConnectionData_GetStringField(agent_info, "mqttBroker.username", &settings->username))
+        {
+            settings->username = NULL;
+            Log_Error("Username field is not specified");
+            goto done;
+        }
+    }
+
+    if (!ADUC_AgentInfo_ConnectionData_GetStringField(agent_info, "mqttBroker.password", &settings->password))
```

Comment:
again, secret shouldn't be in memory for long time to lessen chance ending up in crash dump.  Let's remove password field if not needed.